### PR TITLE
feat: make arrays, plain objects, and those marked live

### PR DESCRIPTION
**Pull request by jdalton**
feat: make arrays, plain objects, and those marked live

Tests are being worked on but getting the initial implementation out for review/discussion.

### CODE REVIEW

**Comment by jdalton** [comment]
Tests are being worked on but getting the initial implementation out for review/discussion.

File: packages/near-membrane-base/src/__tests__/environment.spec.ts [hunk 1]
```diff
@@ -130,7 +130,7 @@ describe('VirtualEnvironment', () => {
     });
 
     describe('VirtualEnvironment.prototype.remap', () => {
-        it('Skips index-like properties', () => {
+        it('does not skip index-like properties', () => {
             // Ignoring "Property 'assertions' does not exist on type '{...}'."
             // @ts-ignore
             expect.assertions(1);
```
Comment (jdalton) [inline packages/near-membrane-base/src/__tests__/environment.spec.ts:136]: Before we had no logic to actually skip index-like properties in the membrane. It relied on the fact that mutations where allowed to some objects that weren't live. My hunch is  that we in this PR we allow an object whose prototype's prototype is null to be considered as "plain" (using ObservableMemebrane's implementation) and it hits on something with that.

File: packages/near-membrane-base/src/membrane.ts [hunk 2]
```diff
@@ -725,7 +768,9 @@ export function createMembraneMarshall() {
                 if (foreignTargetTraits & TargetTraits.Revoked) {
                     revoke();
                 }
-                if (!trapMutations) {
+                if (foreignTargetTraits & TargetTraits.IsArray) {
+                    this.makeProxyLive();
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:772]: But this should only really happen if `trapMutations` is true, otherwise for blue this line might get called when it is already the case that the default hooks are live (in the case of the blue side for example)

**Review by caridy — CHANGES_REQUESTED** [review]

File: packages/near-membrane-base/src/membrane.ts [hunk 3]
```diff
@@ -611,6 +644,44 @@ export function createMembraneMarshall() {
             return fn;
         }
 
+        function isLiveObject(value: any): boolean {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:647]: We need to figure a way for this to be called for valid proxy targets that are non-arrays, so we can avoid all the conditions below.

File: packages/near-membrane-base/src/membrane.ts [hunk 4]
```diff
@@ -611,6 +644,44 @@ export function createMembraneMarshall() {
...
+                    LOCKER_LIVE_INSTANCE_MARKER_SYMBOL,
+                ]) &&
+                ReflectHas(value, LOCKER_LIVE_INSTANCE_MARKER_SYMBOL)
+            ) {
+                return true;
+            }
+            return false;
+        }
+
+        function isPlainObject(value: any): boolean {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:675]: let's fold that into `isLiveObject` so we can optimize it better, and no one else should call this abstract block.

File: packages/near-membrane-base/src/membrane.ts (hunk 2 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: You're right. Blue is different than live.

Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Fixed.

File: packages/near-membrane-base/src/membrane.ts [hunk 5]
```diff
@@ -611,6 +644,44 @@ export function createMembraneMarshall() {
...
 
+        function isLiveObject(value: any): boolean {
+            if (value === null || value === undefined) {
+                return false;
+            }
+            if (typeof value === 'object') {
+                if (
+                    // Arrays are marked as live in the BoundaryProxyHandler constructor.
+                    // isArrayOrNotOrThrowForRevoked(value) ||
+                    ArrayBufferIsView(value) ||
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:655]: This can be called with anything... objects, functions, etc... and it is pretty fast.... so maybe it should be alone in the check for quickness. Something like this at the top of the method:

```
if (ArrayBufferIsView(value)) { return true; }
```

**Review by caridy — CHANGES_REQUESTED** [review]

File: packages/near-membrane-base/src/membrane.ts [hunk 6]
```diff
@@ -611,6 +644,44 @@ export function createMembraneMarshall() {
...
+                    isPlainObject(value)
+                ) {
+                    return true;
+                }
+            }
+            if (ReflectApply(ObjectProtoHasOwnProperty, value, [LOCKER_LIVE_VALUE_MARKER_SYMBOL])) {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:661]: this should be the less common case, probably should be last on this method.

File: packages/near-membrane-base/src/membrane.ts (hunk 4 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Done!

File: packages/near-membrane-base/src/membrane.ts [hunk 7]
```diff
@@ -611,6 +644,43 @@ export function createMembraneMarshall() {
...
+                return true;
+            }
+            if (
+                !ReflectApply(ObjectProtoHasOwnProperty, value, [
+                    LOCKER_LIVE_INSTANCE_MARKER_SYMBOL,
+                ]) &&
+                ReflectHas(value, LOCKER_LIVE_INSTANCE_MARKER_SYMBOL)
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:677]: I think the idea here is to get the proto, then check on the class instance symbol on the proto itself. has will walk all the way up... that seems problematic.

File: packages/near-membrane-base/src/membrane.ts (hunk 6 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Done.

File: packages/near-membrane-base/src/membrane.ts [hunk 8]
```diff
@@ -611,6 +644,42 @@ export function createMembraneMarshall() {
...
+                    proto === ObjectProto ||
+                    proto === null ||
+                    ReflectGetPrototypeOf(proto) === null
+                ) {
+                    return true;
+                }
+            }
+            if (
+                !ReflectApply(ObjectProtoHasOwnProperty, value, [
+                    LOCKER_LIVE_INSTANCE_MARKER_SYMBOL,
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:672]: this apply is done twice in this method, it is always false at this point.

**Review by caridy — CHANGES_REQUESTED** [review]

File: packages/near-membrane-base/src/membrane.ts (hunk 7 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:674]: Should we walk prototypes checking?  So checking prototype's of prototypes too?

File: packages/near-membrane-base/src/membrane.ts (hunk 5 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Done

File: packages/near-membrane-base/src/membrane.ts (hunk 8 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:672]: The `ObjectProtoHasOwnProperty` checks are for two different properties. The follow-up check for `LOCKER_LIVE_INSTANCE_MARKER_SYMBOL` is to make sure it isn't an own property but rather inherited.

**Comment by jdalton** [comment]
Pulling out the `LOCKER_LIVE_INSTANCE_MARKER_SYMBOL` stuff for now to focus on the immediate issue of liveness for arrays, views, and plain objects (that aren't prototype objects).

File: packages/near-membrane-base/src/membrane.ts (hunk 3 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:647]: Worked with @dgitin to make the checks less heavy.

File: packages/near-membrane-base/src/membrane.ts [hunk 9]
```diff
@@ -611,6 +644,34 @@ export function createMembraneMarshall() {
...
         }
 
+        function isLiveObject(value: any): boolean {
+            if (value === null || value === undefined || value === ObjectProto) {
+                return false;
+            }
+            if (typeof value === 'object') {
+                if (
+                    // Arrays are marked as live in the BoundaryProxyHandler constructor.
+                    // isArrayOrNotOrThrowForRevoked(value) ||
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:654]: this is not related I believe, let's remove that... ArrayBugger.isView will never throw.

File: packages/near-membrane-base/src/membrane.ts [hunk 10]
```diff
@@ -611,6 +644,34 @@ export function createMembraneMarshall() {
...
+                    return true;
+                }
+                const { constructor } = value;
+                if (constructor === ObjectCtor) {
+                    // If the constructor, own or inherited, points to `Object`
+                    // then `value` is not likely a prototype object.
+                    return true;
+                }
+                if (ReflectGetPrototypeOf(value) === null) {
+                    // Ensure `value` is not an `Object.prototype` from an iframe.
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:666]: why do we care about this? if it is the `Object.prototype` value from any realm it should never be mutate.

File: packages/near-membrane-base/src/membrane.ts [hunk 11]
```diff
@@ -611,6 +644,34 @@ export function createMembraneMarshall() {
             return fn;
         }
 
+        function isLiveObject(value: any): boolean {
+            if (value === null || value === undefined || value === ObjectProto) {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:648]:
```suggestion
            if (value === ObjectProto) {
```

the value should not be type any, it should be a proxy target type, this should never be called with a thing that is not a valid proxy target.

File: packages/near-membrane-base/src/membrane.ts [hunk 12]
```diff
@@ -611,6 +644,34 @@ export function createMembraneMarshall() {
...
+                return false;
+            }
+            if (typeof value === 'object') {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:651]: is this check really needed? I don't think so... if it is just to differentiate functions from objects, then ok.

File: packages/near-membrane-base/src/membrane.ts (hunk 9 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:654]: It's a comment to show that we don't need to do the array check. It's not code to guard anything.

File: packages/near-membrane-base/src/membrane.ts (hunk 10 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:666]: `Object.prototype` has a `ReflectGetPrototypeOf(value) === null`. So if we don't guard for it then it'll pass and be mutable.

File: packages/near-membrane-base/src/membrane.ts (hunk 11 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Fixed

File: packages/near-membrane-base/src/membrane.ts (hunk 9 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: Removed and updated comment to avoid confusion.

File: packages/near-membrane-base/src/membrane.ts (hunk 12 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts]: the target could be a function.

File: packages/near-membrane-base/src/membrane.ts [hunk 13]
```diff
@@ -610,6 +643,34 @@ export function createMembraneMarshall() {
             return fn;
         }
 
+        function isLiveTarget(target: ProxyTarget): boolean {
+            if (target === ObjectProto) {
```
Comment (caridy) [inline packages/near-membrane-base/src/membrane.ts:647]: thinking more about this... Object.prototype is reflective in the membrane, so target cannot be ObjectProto, we can remove this condition all together.

**Review by caridy — APPROVED** [review]
LGTM

File: packages/near-membrane-base/src/membrane.ts (hunk 13 above)
Comment (jdalton) [reply packages/near-membrane-base/src/membrane.ts:647]: Since these are handlers on both sides colors. I wonder if it could be the primary at some point (reflective).
